src/external.rs

```diff
@@ -0,0 +1,263 @@
+        let mut highest_res: u64 = 0;
+        let mut webm_highest_res: u64 = 0;
+        for thumbnail in thumbnails {
+            if let Some(Value::Number(height)) = thumbnail.get("height") {
+                if let Some(Value::String(url)) = thumbnail.get("url") {
+                    let height = height.as_u64().unwrap();
+                    if url.contains("webm") {
+                        if webm_highest_res < height {
+                            webm_highest_res = height;
+                            webm_max_res_url = Some(url);
+                        }
+                    } else if highest_res < height {
+                        highest_res = height;
+                        max_res_url = Some(url);
+                    }
+                }
+            }
+        }
+        if highest_res >= webm_highest_res {
+            max_res_url.map(|s| s.into())
+        } else {
+            webm_max_res_url.map(|s| s.into())
+        }
+    } else {
+        None
+    }
+}
+
+fn get_stream_title(data: &Value) -> Option<String> {
+    if let Some(Value::String(nym)) = data.get("fulltitle") {
+        Some(nym.to_owned())
+    } else if let Some(Value::String(nym)) = data.get("title") {
+        Some(nym.to_owned())
+    } else {
+        None
+    }
+}
+
+pub async fn download_stream(url_hash: UrlHash, req_client: Client) {
+    let base_path = format!["{STREAM_DIR}/{}", url_hash.0];
+    let json_path = format!["{base_path}/info.json"];
+    let data: Value = from_reader(File::open(json_path).unwrap()).unwrap();
+    let audio_url = get_audio_url(&data);
+    let video_url = get_video_url(&data);
+    let subs_urls = get_subs_urls(&data);
+    let thumbnail_url = get_max_res_thumbnail_url(&data);
+    let title = get_stream_title(&data).unwrap_or("unkown".into());
+
+    let audio_data = if let Some((url, extension)) = audio_url {
+        if let Ok(res) = req_client.post(url).send().await {
+            Some((res.bytes(), extension))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+    let video_data = if let Some((url, extension)) = video_url {
+        if let Ok(res) = req_client.post(url).send().await {
+            Some((res.bytes(), extension))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+    let mut subs_data = Vec::with_capacity(subs_urls.len());
+    for (lang, _name, url, extension) in subs_urls {
+        if let Ok(res) = req_client.post(url).send().await {
+            subs_data.push((res.bytes(), lang, extension));
+        }
+    }
+    let thumbnail_data = if let Some(url) = thumbnail_url {
+        let extension = url
+            .rsplit('.')
+            .take(1)
+            .map(|s| s.to_owned())
+            .next()
+            .unwrap_or_default();
+        if let Ok(res) = req_client.post(url).send().await {
+            Some((res.bytes(), extension))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
```

Comment:
Ideally all these downloads should happen in parallel

Reply:
Should be fixed in new changes.